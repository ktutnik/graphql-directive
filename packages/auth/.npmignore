test
src
coverage
.npmignore
tsconfig.json
jest.config.debug.ts 
jest.config.ts 
package-lock.json
yarn-error.log
yarn.lock