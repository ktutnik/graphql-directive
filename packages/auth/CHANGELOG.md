# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 1.0.0 (2023-04-07)

### Features

- Add authorization directive ([#33](https://github.com/ktutnik/graphql-directive/issues/33)) ([a46d4a2](https://github.com/ktutnik/graphql-directive/commit/a46d4a2211ad9117b33eaeffb0670d87167c7c24))
