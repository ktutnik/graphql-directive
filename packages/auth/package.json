{
  "name": "@graphql-directive/auth",
  "version": "1.0.1",
  "description": "GraphQL authorization directive",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "scripts": {
    "test": "jest",
    "build": "tsc",
    "clean": "trash lib coverage"
  },
  "keywords": [
    "graphql",
    "auth",
    "authorize",
    "authorization",
    "directive",
    "graphql-directive"
  ],
  "author": "I Ketut Sandiarsa",
  "license": "MIT",
  "peerDependencies": {
    "graphql": "^14.0.0 || ^15.0.0 || ^16.0.0 || ^17.0.0"
  },
  "dependencies": {
    "@graphql-directive/core": "^1.0.1"
  },
  "funding": [
    "https://github.com/sponsors/ktutnik"
  ],
  "devDependencies": {
    "@graphql-tools/schema": "^9.0.17",
    "@types/jest": "^29.5.0",
    "graphql": "^16.6.0",
    "jest": "^29.5.0",
    "trash-cli": "^5.0.0",
    "ts-jest": "^29.1.0",
    "ts-node": "^10.9.1"
  },
  "bugs": {
    "url": "https://github.com/ktutnik/graphql-directive/issues"
  },
  "repository": {
    "url": "https://github.com/ktutnik/graphql-directive"
  },
  "publishConfig": {
    "access": "public"
  }
}
