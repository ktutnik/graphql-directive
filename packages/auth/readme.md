# GraphQL Authorization Directive

[![Node.js CI](https://github.com/ktutnik/graphql-directive/actions/workflows/test.yml/badge.svg)](https://github.com/ktutnik/graphql-directive/actions/workflows/test.yml)
[![Coverage Status](https://coveralls.io/repos/github/ktutnik/graphql-directive/badge.svg?branch=master)](https://coveralls.io/github/ktutnik/graphql-directive?branch=master)

A TypeScript/JavaScript library that provides an easy way to add authorization logic to your Node.js GraphQL API using directives. 

## Motivation
GraphQL Authorization Directive aims to simplify the process of adding authorization to your GraphQL API. By using directives, you can easily apply authorization logic to your schema without having to manually implement complex middleware functions.

## Example Usage
```javascript
import auth from "@graphql-directive/auth"
import { ApolloServer } from "@apollo/server"
import { startStandaloneServer } from "@apollo/server/standalone"
import { makeExecutableSchema } from "@graphql-tools/schema"

const typeDefs = `
    type User {
        name: String!
        email: String!
        role: String @authorize(policy: "admin")
    }
    input UserInput {
        name: String!
        email: String! 
        role: String @authorize(policy: "admin")
    }
    type Query {
        getUsers: [User]!
    }
    type Mutation {
        addUser(user:UserInput!): Boolean!
        modifyUser(user: UserInput!): Boolean!
    }
`

const transform = auth.createTransformer({
    policies: {
        admin: ({ contextValue }) => contextValue.user.role === "admin",
        isLogin: ({ contextValue }) => !!contextValue.user
    }
})

const schema = transform(makeExecutableSchema({
    typeDefs: [auth.typeDefs, typeDefs],
    resolvers: {
        Query: { getUsers: () => ([]) },
        Mutation: {
            addUser: () => true,
            modifyUser: ()=> true
        }
    }
}))

const server = new ApolloServer({ schema })
startStandaloneServer(server, { context: async ({ req, res }) => ({}) }).then(x => console.log(x.url))
```

## API Documentation
