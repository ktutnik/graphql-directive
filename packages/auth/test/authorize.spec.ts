import auth from "../src"
import { makeExecutableSchema } from "@graphql-tools/schema"
import { GraphQLSchema, graphql } from "graphql"

function createSchema(def: string, resolver: any = { Mutation: { test: () => true } }, queryResolution: "ThrowError" | "Filter" = "ThrowError") {
    const typeDefs = /* GraphQL */ `
        type Query { name: String }
    `
    const transform = auth.createTransformer({
        policies: {
            admin: ({ contextValue }) => contextValue?.user?.role === "admin",
            user: ({ contextValue }) => contextValue?.user?.role === "user",
            authenticated: ({ contextValue }) => !!contextValue?.user
        },
        queryResolution
    })
    return transform(makeExecutableSchema({
        typeDefs: [auth.typeDefs, typeDefs, def],
        resolvers: {
            ...resolver
        }
    }))
}

describe("Mutation authorization", () => {
    describe("On field argument", () => {
        const test = (schema: GraphQLSchema, role: string, source?: string) => {
            return graphql({
                schema, source: source ?? `mutation { test(name: "Wayan Koster", role: "admin") }`,
                contextValue: { user: { role } }
            })
        }
        it("Should authorize with proper role", async () => {
            const schema = createSchema(`
                type Mutation { 
                    test(name:String!, role: String @authorize(policy: "admin")): Boolean!
                }`)
            expect((await test(schema, "admin")).data!.test).toBe(true)
            expect((await test(schema, "user")).errors![0].extensions).toMatchSnapshot()
        })

        it("Should allow unauthorized user when not providing value", async () => {
            const schema = createSchema(`
                type Mutation { 
                    test(name:String!, role: String @authorize(policy: "admin")): Boolean!
                }`)
            expect((await test(schema, "admin", `mutation { test(name: "Wayan Koster") }`)).data!.test).toBe(true)
        })

        it("Should work on array type", async () => {
            const schema = createSchema(`
                type Mutation { 
                    test(name:String!, role: [String] @authorize(policy: "admin")): Boolean!
                }`)
            const source = `mutation { test(name: "Wayan Koster", role: ["admin", "user"]) }`
            expect((await test(schema, "admin", source)).data!.test).toBe(true)
            expect((await test(schema, "user", source)).errors![0].extensions).toMatchSnapshot()
        })

        it("Should able to apply multiple directives", async () => {
            const schema = createSchema(`
                type Mutation { 
                    test(name:String!, role: String @authorize(policy: "admin, user")): Boolean!
                }`)
            expect((await test(schema, "admin")).data!.test).toBe(true)
            expect((await test(schema, "user")).data!.test).toBe(true)
            expect((await test(schema, "authenticated")).errors![0].extensions).toMatchSnapshot()
        })

        it("Should throw error when specify invalid policy name", async () => {
            const schema = createSchema(`
                type Mutation { 
                    test(name:String!, role: String @authorize(policy: "superadmin")): Boolean!
                }`)
            expect((await test(schema, "user")).errors![0]).toMatchSnapshot()
        })
    })

    describe("On mutation field", () => {
        const test = (schema: GraphQLSchema, role: string, source?: string) => {
            return graphql({
                schema, source: source ?? `mutation { test(name: "Wayan Koster", role: "admin") }`,
                contextValue: { user: { role } }
            })
        }
        it("Should authorize with proper role", async () => {
            const schema = createSchema(`
                type Mutation { 
                    test(name:String!, role: String): Boolean! @authorize(policy: "admin")
                }`)
            expect((await test(schema, "admin")).data!.test).toBe(true)
            expect((await test(schema, "user")).errors![0].extensions).toMatchSnapshot()
        })

        it("Should able to apply multiple directives", async () => {
            const schema = createSchema(`
                type Mutation { 
                    test(name:String!, role: String): Boolean! @authorize(policy: "admin, user")
                }`)
            expect((await test(schema, "admin")).data!.test).toBe(true)
            expect((await test(schema, "user")).data!.test).toBe(true)
            expect((await test(schema, "authenticated")).errors![0].extensions).toMatchSnapshot()
        })

        it("Should throw error when specify invalid policy name", async () => {
            const schema = createSchema(`
                type Mutation { 
                    test(name:String!, role: String): Boolean! @authorize(policy: "superadmin")
                }`)
            expect((await test(schema, "admin")).errors![0]).toMatchSnapshot()
        })
    })

    describe("On input type field", () => {
        const test = (schema: GraphQLSchema, role: string, source?: string) => {
            return graphql({
                schema, source: source ?? `mutation { test(user: { name: "Wayan Koster", role: "admin" }) }`,
                contextValue: { user: { role } }
            })
        }
        it("Should authorize with proper role", async () => {
            const schema = createSchema(`
                input User {
                    name:String!
                    role:String @authorize(policy: "admin")
                }
                type Mutation { 
                    test(user:User!): Boolean!
                }`)

            expect((await test(schema, "admin")).data!.test).toBe(true)
            expect((await test(schema, "user")).errors![0].extensions).toMatchSnapshot()
        })

        it("Should pass if provided no value", async () => {
            const schema = createSchema(`
                input User {
                    name:String!
                    role:String @authorize(policy: "admin")
                }
                type Mutation { 
                    test(user:User!): Boolean!
                }`)
            expect((await test(schema, "user", `mutation { test(user: { name: "Wayan Koster" }) }`)).data?.test).toBe(true)
        })

        it("Should able to apply multiple directives", async () => {
            const schema = createSchema(`
                input User {
                    name:String!
                    role:String @authorize(policy: "admin, user") 
                }
                type Mutation { 
                    test(user:User!): Boolean!
                }`)
            expect((await test(schema, "admin")).data!.test).toBe(true)
            expect((await test(schema, "user")).data!.test).toBe(true)
            expect((await test(schema, "authenticated")).errors![0].extensions).toMatchSnapshot()
        })

        it("Should able to apply in array type", async () => {
            const schema = createSchema(`
                input User {
                    name:String!
                    role:String @authorize(policy: "admin") 
                }
                type Mutation { 
                    test(user:[User]!): Boolean!
                }`)
            const source = `mutation { test(user: [{ name: "Wayan Koster", role: "admin" }, { name: "Wayan Koster", role: "admin" }]) }`
            const sourceVariation = `mutation { test(user: [{ name: "Wayan Koster", role: "admin" }, { name: "Wayan Koster" }]) }`
            expect((await test(schema, "admin", source)).data!.test).toBe(true)
            expect((await test(schema, "authenticated", source)).errors![0].extensions).toMatchSnapshot()
            expect((await test(schema, "authenticated", sourceVariation)).errors![0].extensions).toMatchSnapshot()
        })

        it("Should throw error when specify invalid policy name", async () => {
            const schema = createSchema(`
                input User {
                    name:String!
                    role:String @authorize(policy: "superadmin")
                }
                type Mutation { 
                    test(user:User!): Boolean!
                }`)

            expect((await test(schema, "admin")).errors![0]).toMatchSnapshot()
        })
    })
})

describe("Query authorization", () => {
    const test = (schema: GraphQLSchema, role: string) => {
        return graphql({
            schema, source: `query { test(name: "Wayan Koster", role: "admin") }`,
            contextValue: { user: { role } }
        })
    }
    const resolver = { Query: { test: () => true } }

    describe("Applied on query field argument", () => {
        it("Should authorize with proper role", async () => {
            const schema = createSchema(`
                extend type Query { 
                    test(name:String!, role: String @authorize(policy: "admin")): Boolean!
                }`, resolver)
            expect((await test(schema, "admin")).data!.test).toBe(true)
            expect((await test(schema, "user")).errors![0].extensions).toMatchSnapshot()
        })

        it("Should able to apply multiple directives", async () => {
            const schema = createSchema(`
                extend type Query { 
                    test(name:String!, role: String @authorize(policy: "admin, user")): Boolean!
                }`, resolver)
            expect((await test(schema, "admin")).data!.test).toBe(true)
            expect((await test(schema, "user")).data!.test).toBe(true)
            expect((await test(schema, "authenticated")).errors![0].extensions).toMatchSnapshot()
        })
        it("Should throw error when specify invalid policy name", async () => {
            const schema = createSchema(`
                extend type Query { 
                    test(name:String!, role: String @authorize(policy: "superadmin")): Boolean!
                }`, resolver)
            expect((await test(schema, "admin")).errors![0]).toMatchSnapshot()
        })
    })
})

describe("Query authorization ThrowError mode", () => {

    describe("Applied on query field with primitive type", () => {
        const resolver = { Query: { test: () => true } }
        const test = (schema: GraphQLSchema, role: string) => {
            return graphql({
                schema, source: `query { test }`,
                contextValue: { user: { role } }
            })
        }

        it("Should authorize with proper role", async () => {
            const schema = createSchema(`
                extend type Query { 
                    test: Boolean! @authorize(policy: "admin")
                }`, resolver)
            expect((await test(schema, "admin")).data!.test).toBe(true)
            expect((await test(schema, "user")).errors![0].extensions).toMatchSnapshot()
        })

        it("Should authorize with multiple directives", async () => {
            const schema = createSchema(`
                extend type Query { 
                    test: Boolean! @authorize(policy: "admin, user")
                }`, resolver)
            expect((await test(schema, "admin")).data!.test).toBe(true)
            expect((await test(schema, "user")).data!.test).toBe(true)
            expect((await test(schema, "authenticated")).errors![0].extensions).toMatchSnapshot()
        })
    })

    describe("Applied on complex type property", () => {
        const resolver = {
            Query: {
                test: () => ({ name: "Wayan Koster", role: "admin" })
            }
        }
        const test = (schema: GraphQLSchema, role: string) => {
            return graphql({
                schema, source: `query { test { name, role } }`,
                contextValue: { user: { role } }
            })
        }

        it("Should authorize with proper role", async () => {
            const schema = createSchema(`
                type User {
                    name:String!
                    role:String! @authorize(policy: "admin")
                }
                extend type Query { 
                    test: User!
                }`, resolver)
            expect((await test(schema, "admin")).data!.test).toMatchSnapshot()
            expect((await test(schema, "user")).errors![0].extensions).toMatchSnapshot()
        })

        it("Should authorize with multiple directives", async () => {
            const schema = createSchema(`
                type User {
                    name:String!
                    role:String! @authorize(policy: "admin, user")
                }
                extend type Query { 
                    test: User!
                }`, resolver)
            expect((await test(schema, "admin")).data!.test).toMatchSnapshot()
            expect((await test(schema, "user")).data!.test).toMatchSnapshot()
            expect((await test(schema, "authenticated")).errors![0].extensions).toMatchSnapshot()
        })
    })

    describe("Applied on nested complex type property", () => {
        const resolver = {
            Query: {
                test: () => ({ name: "Wayan Koster", role: "admin", partner: { name: "Wayan Koster", role: "admin" } })
            }
        }
        const test = (schema: GraphQLSchema, role: string) => {
            return graphql({
                schema, source: `query { test { name, partner { name, role } } }`,
                contextValue: { user: { role } }
            })
        }

        it("Should authorize with proper role", async () => {
            const schema = createSchema(`
                type User {
                    name:String!
                    role:String! @authorize(policy: "admin")
                    partner: User!
                }
                extend type Query { 
                    test: User!
                }`, resolver)
            expect((await test(schema, "admin")).data!.test).toMatchSnapshot()
            expect((await test(schema, "user")).errors![0].extensions).toMatchSnapshot()
        })

        it("Should authorize with multiple directives", async () => {
            const schema = createSchema(`
                type User {
                    name:String!
                    role:String! @authorize(policy: "admin, user")
                    partner: User!
                }
                extend type Query { 
                    test: User!
                }`, resolver)
            expect((await test(schema, "admin")).data!.test).toMatchSnapshot()
            expect((await test(schema, "user")).data!.test).toMatchSnapshot()
            expect((await test(schema, "authenticated")).errors![0].extensions).toMatchSnapshot()
        })
    })

    describe("Applied on array of complex type property", () => {
        const resolver = {
            Query: {
                test: () => ([
                    { name: "Wayan Koster", role: "admin" },
                    { name: "Wayan Koster", role: "admin" }
                ])
            }
        }
        const test = (schema: GraphQLSchema, role: string) => {
            return graphql({
                schema, source: `query { test { name, role } }`,
                contextValue: { user: { role } }
            })
        }

        it("Should authorize with proper role", async () => {
            const schema = createSchema(`
                type User {
                    name:String!
                    role:String! @authorize(policy: "admin")
                }
                extend type Query { 
                    test: [User]!
                }`, resolver)
            expect((await test(schema, "admin")).data!.test).toMatchSnapshot()
            expect((await test(schema, "user")).errors![0].extensions).toMatchSnapshot()
        })

        it("Should authorize with multiple directives", async () => {
            const schema = createSchema(`
                type User {
                    name:String!
                    role:String! @authorize(policy: "admin, user") 
                }
                extend type Query { 
                    test: [User]!
                }`, resolver)
            expect((await test(schema, "admin")).data!.test).toMatchSnapshot()
            expect((await test(schema, "user")).data!.test).toMatchSnapshot()
            expect((await test(schema, "authenticated")).errors![0].extensions).toMatchSnapshot()
        })
    })
})


describe("Query authorization Filter mode", () => {

    describe("Applied on query field with primitive type", () => {
        const resolver = { Query: { test: () => true } }
        const test = (schema: GraphQLSchema, role: string) => {
            return graphql({
                schema, source: `query { test }`,
                contextValue: { user: { role } }
            })
        }

        it("Should authorize with proper role", async () => {
            const schema = createSchema(`
                extend type Query { 
                    test: Boolean @authorize(policy: "admin")
                }`, resolver, "Filter")
            expect((await test(schema, "admin")).data!.test).toBe(true)
            expect((await test(schema, "user")).data!.test).toBeNull()
        })

        it("Should throw error when provided non null type", async () => {
            expect(() => createSchema(`
                extend type Query { 
                    test: Boolean! @authorize(policy: "admin")
                }`, resolver, "Filter")).toThrowErrorMatchingSnapshot()
        })

        it("Should authorize with multiple directives", async () => {
            const schema = createSchema(`
                extend type Query { 
                    test: Boolean @authorize(policy: "admin, user") 
                }`, resolver, "Filter")
            expect((await test(schema, "admin")).data!.test).toBe(true)
            expect((await test(schema, "user")).data!.test).toBe(true)
            expect((await test(schema, "authenticated")).data!.test).toBeNull()
        })
    })

    describe("Applied on complex type property", () => {
        const resolver = {
            Query: {
                test: () => ({ name: "Wayan Koster", role: "admin" })
            }
        }
        const test = (schema: GraphQLSchema, role: string) => {
            return graphql({
                schema, source: `query { test { name, role } }`,
                contextValue: { user: { role } }
            })
        }

        it("Should authorize with proper role", async () => {
            const schema = createSchema(`
                type User {
                    name:String!
                    role:String @authorize(policy: "admin")
                }
                extend type Query { 
                    test: User!
                }`, resolver, "Filter")
            expect((await test(schema, "admin")).data!.test).toMatchSnapshot()
            expect((await test(schema, "user")).data!.test).toMatchSnapshot()
        })

        it("Should throw error when applied on non nullable type", async () => {
            expect(() => createSchema(`
                type User {
                    name:String!
                    role:String! @authorize(policy: "admin")
                }
                extend type Query { 
                    test: User!
                }`, resolver, "Filter")).toThrowErrorMatchingSnapshot()
        })

        it("Should authorize with multiple directives", async () => {
            const schema = createSchema(`
                type User {
                    name:String!
                    role:String @authorize(policy: "admin, user") 
                }
                extend type Query { 
                    test: User!
                }`, resolver, "Filter")
            expect((await test(schema, "admin")).data!.test).toMatchSnapshot()
            expect((await test(schema, "user")).data!.test).toMatchSnapshot()
            expect((await test(schema, "authenticated")).data!.test).toMatchSnapshot()
        })
    })

    describe("Applied on nested complex type property", () => {
        const resolver = {
            Query: {
                test: () => ({ name: "Wayan Koster", role: "admin", partner: { name: "Wayan Koster", role: "admin" } })
            }
        }
        const test = (schema: GraphQLSchema, role: string) => {
            return graphql({
                schema, source: `query { test { name, role, partner { name, role } } }`,
                contextValue: { user: { role } }
            })
        }

        it("Should authorize with proper role", async () => {
            const schema = createSchema(`
                type User {
                    name:String!
                    role:String @authorize(policy: "admin")
                    partner: User!
                }
                extend type Query { 
                    test: User!
                }`, resolver, "Filter")
            expect((await test(schema, "admin")).data!.test).toMatchSnapshot()
            expect((await test(schema, "user")).data!.test).toMatchSnapshot()
        })

        it("Should throw error when applied on non nullable type", async () => {
            expect(() => createSchema(`
                type User {
                    name:String!
                    role:String! @authorize(policy: "admin")
                    partner: User!
                }
                extend type Query { 
                    test: User!
                }`, resolver, "Filter")).toThrowErrorMatchingSnapshot()
        })

        it("Should authorize with multiple directives", async () => {
            const schema = createSchema(`
                type User {
                    name:String!
                    role:String @authorize(policy: "admin, user") 
                    partner: User!
                }
                extend type Query { 
                    test: User!
                }`, resolver, "Filter")
            expect((await test(schema, "admin")).data!.test).toMatchSnapshot()
            expect((await test(schema, "user")).data!.test).toMatchSnapshot()
            expect((await test(schema, "authenticated")).data!.test).toMatchSnapshot()
        })
    })

    describe("Applied on array of complex type property", () => {
        const resolver = {
            Query: {
                test: () => ([
                    { name: "Wayan Koster", role: "admin" },
                    { name: "Wayan Koster", role: "admin" }
                ])
            }
        }
        const test = (schema: GraphQLSchema, role: string) => {
            return graphql({
                schema, source: `query { test { name, role } }`,
                contextValue: { user: { role } }
            })
        }

        it("Should authorize with proper role", async () => {
            const schema = createSchema(`
                type User {
                    name:String!
                    role:String @authorize(policy: "admin")
                }
                extend type Query { 
                    test: [User]!
                }`, resolver, "Filter")
            expect((await test(schema, "admin")).data!.test).toMatchSnapshot()
            expect((await test(schema, "user")).data!.test).toMatchSnapshot()
        })

        it("Should throw error when applied on non nullable type", async () => {
            expect(() => createSchema(`
                type User {
                    name:String!
                    role:String! @authorize(policy: "admin")
                }
                extend type Query { 
                    test: [User]!
                }`, resolver, "Filter")).toThrowErrorMatchingSnapshot()
        })

        it("Should authorize with multiple directives", async () => {
            const schema = createSchema(`
                type User {
                    name:String!
                    role:String @authorize(policy: "admin, user") 
                }
                extend type Query { 
                    test: [User]!
                }`, resolver, "Filter")
            expect((await test(schema, "admin")).data!.test).toMatchSnapshot()
            expect((await test(schema, "user")).data!.test).toMatchSnapshot()
            expect((await test(schema, "authenticated")).data!.test).toMatchSnapshot()
        })
    })
})