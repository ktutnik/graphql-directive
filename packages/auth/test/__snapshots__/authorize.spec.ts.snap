// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation authorization On complex return type (Filter mode) Should authorize return type with proper role 1`] = `
{
  "name": "Wayan Koster",
  "role": "admin",
}
`;

exports[`Mutation authorization On complex return type (Filter mode) Should authorize return type with proper role 2`] = `
{
  "name": "Wayan Koster",
  "role": null,
}
`;

exports[`Mutation authorization On complex return type (ThrowError mode) Should authorize return type with proper role 1`] = `
{
  "name": "Wayan Koster",
  "role": "admin",
}
`;

exports[`Mutation authorization On complex return type (ThrowError mode) Should authorize return type with proper role 2`] = `
{
  "code": "GRAPHQL_AUTHORIZATION_FAILED",
  "paths": [
    "test.role",
  ],
}
`;

exports[`Mutation authorization On field argument Should able to apply multiple directives 1`] = `
{
  "code": "GRAPHQL_AUTHORIZATION_FAILED",
  "paths": [
    "test.role",
  ],
}
`;

exports[`Mutation authorization On field argument Should authorize with proper role 1`] = `
{
  "code": "GRAPHQL_AUTHORIZATION_FAILED",
  "paths": [
    "test.role",
  ],
}
`;

exports[`Mutation authorization On field argument Should throw error when specify invalid policy name 1`] = `[GraphQLError: Unknown policy "superadmin" on test.role]`;

exports[`Mutation authorization On field argument Should work on array type 1`] = `
{
  "code": "GRAPHQL_AUTHORIZATION_FAILED",
  "paths": [
    "test.role.0",
    "test.role.1",
  ],
}
`;

exports[`Mutation authorization On input type field Should able to apply in array type 1`] = `
{
  "code": "GRAPHQL_AUTHORIZATION_FAILED",
  "paths": [
    "test.user.0.role",
    "test.user.1.role",
  ],
}
`;

exports[`Mutation authorization On input type field Should able to apply in array type 2`] = `
{
  "code": "GRAPHQL_AUTHORIZATION_FAILED",
  "paths": [
    "test.user.0.role",
  ],
}
`;

exports[`Mutation authorization On input type field Should able to apply multiple directives 1`] = `
{
  "code": "GRAPHQL_AUTHORIZATION_FAILED",
  "paths": [
    "test.user.role",
  ],
}
`;

exports[`Mutation authorization On input type field Should authorize with proper role 1`] = `
{
  "code": "GRAPHQL_AUTHORIZATION_FAILED",
  "paths": [
    "test.user.role",
  ],
}
`;

exports[`Mutation authorization On input type field Should throw error when specify invalid policy name 1`] = `[GraphQLError: Unknown policy "superadmin" on test.user.role]`;

exports[`Mutation authorization On mutation field Should able to apply multiple directives 1`] = `
{
  "code": "GRAPHQL_AUTHORIZATION_FAILED",
  "paths": [
    "test",
  ],
}
`;

exports[`Mutation authorization On mutation field Should authorize with proper role 1`] = `
{
  "code": "GRAPHQL_AUTHORIZATION_FAILED",
  "paths": [
    "test",
  ],
}
`;

exports[`Mutation authorization On mutation field Should throw error when specify invalid policy name 1`] = `[GraphQLError: Unknown policy "superadmin" on test]`;

exports[`Query authorization Applied on query field argument Should able to apply multiple directives 1`] = `
{
  "code": "GRAPHQL_AUTHORIZATION_FAILED",
  "paths": [
    "test.role",
  ],
}
`;

exports[`Query authorization Applied on query field argument Should authorize with proper role 1`] = `
{
  "code": "GRAPHQL_AUTHORIZATION_FAILED",
  "paths": [
    "test.role",
  ],
}
`;

exports[`Query authorization Applied on query field argument Should throw error when specify invalid policy name 1`] = `[GraphQLError: Unknown policy "superadmin" on test.role]`;

exports[`Query authorization Filter mode Applied on array of complex type property Should authorize with multiple directives 1`] = `
[
  {
    "name": "Wayan Koster",
    "role": "admin",
  },
  {
    "name": "Wayan Koster",
    "role": "admin",
  },
]
`;

exports[`Query authorization Filter mode Applied on array of complex type property Should authorize with multiple directives 2`] = `
[
  {
    "name": "Wayan Koster",
    "role": "admin",
  },
  {
    "name": "Wayan Koster",
    "role": "admin",
  },
]
`;

exports[`Query authorization Filter mode Applied on array of complex type property Should authorize with multiple directives 3`] = `
[
  {
    "name": "Wayan Koster",
    "role": null,
  },
  {
    "name": "Wayan Koster",
    "role": null,
  },
]
`;

exports[`Query authorization Filter mode Applied on array of complex type property Should authorize with proper role 1`] = `
[
  {
    "name": "Wayan Koster",
    "role": "admin",
  },
  {
    "name": "Wayan Koster",
    "role": "admin",
  },
]
`;

exports[`Query authorization Filter mode Applied on array of complex type property Should authorize with proper role 2`] = `
[
  {
    "name": "Wayan Koster",
    "role": null,
  },
  {
    "name": "Wayan Koster",
    "role": null,
  },
]
`;

exports[`Query authorization Filter mode Applied on array of complex type property Should throw error when applied on non nullable type 1`] = `"Nullable data type is required on User.role when Filter resolution enable"`;

exports[`Query authorization Filter mode Applied on complex type property Should authorize with multiple directives 1`] = `
{
  "name": "Wayan Koster",
  "role": "admin",
}
`;

exports[`Query authorization Filter mode Applied on complex type property Should authorize with multiple directives 2`] = `
{
  "name": "Wayan Koster",
  "role": "admin",
}
`;

exports[`Query authorization Filter mode Applied on complex type property Should authorize with multiple directives 3`] = `
{
  "name": "Wayan Koster",
  "role": null,
}
`;

exports[`Query authorization Filter mode Applied on complex type property Should authorize with proper role 1`] = `
{
  "name": "Wayan Koster",
  "role": "admin",
}
`;

exports[`Query authorization Filter mode Applied on complex type property Should authorize with proper role 2`] = `
{
  "name": "Wayan Koster",
  "role": null,
}
`;

exports[`Query authorization Filter mode Applied on complex type property Should throw error when applied on non nullable type 1`] = `"Nullable data type is required on User.role when Filter resolution enable"`;

exports[`Query authorization Filter mode Applied on nested complex type property Should authorize with multiple directives 1`] = `
{
  "name": "Wayan Koster",
  "partner": {
    "name": "Wayan Koster",
    "role": "admin",
  },
  "role": "admin",
}
`;

exports[`Query authorization Filter mode Applied on nested complex type property Should authorize with multiple directives 2`] = `
{
  "name": "Wayan Koster",
  "partner": {
    "name": "Wayan Koster",
    "role": "admin",
  },
  "role": "admin",
}
`;

exports[`Query authorization Filter mode Applied on nested complex type property Should authorize with multiple directives 3`] = `
{
  "name": "Wayan Koster",
  "partner": {
    "name": "Wayan Koster",
    "role": null,
  },
  "role": null,
}
`;

exports[`Query authorization Filter mode Applied on nested complex type property Should authorize with proper role 1`] = `
{
  "name": "Wayan Koster",
  "partner": {
    "name": "Wayan Koster",
    "role": "admin",
  },
  "role": "admin",
}
`;

exports[`Query authorization Filter mode Applied on nested complex type property Should authorize with proper role 2`] = `
{
  "name": "Wayan Koster",
  "partner": {
    "name": "Wayan Koster",
    "role": null,
  },
  "role": null,
}
`;

exports[`Query authorization Filter mode Applied on nested complex type property Should throw error when applied on non nullable type 1`] = `"Nullable data type is required on User.role when Filter resolution enable"`;

exports[`Query authorization Filter mode Applied on query field with primitive type Should throw error when provided non null type 1`] = `"Nullable data type is required on Query.test when Filter resolution enable"`;

exports[`Query authorization ThrowError mode Applied on array of complex type property Should authorize with multiple directives 1`] = `
[
  {
    "name": "Wayan Koster",
    "role": "admin",
  },
  {
    "name": "Wayan Koster",
    "role": "admin",
  },
]
`;

exports[`Query authorization ThrowError mode Applied on array of complex type property Should authorize with multiple directives 2`] = `
[
  {
    "name": "Wayan Koster",
    "role": "admin",
  },
  {
    "name": "Wayan Koster",
    "role": "admin",
  },
]
`;

exports[`Query authorization ThrowError mode Applied on array of complex type property Should authorize with multiple directives 3`] = `
{
  "code": "GRAPHQL_AUTHORIZATION_FAILED",
  "paths": [
    "test.role",
  ],
}
`;

exports[`Query authorization ThrowError mode Applied on array of complex type property Should authorize with proper role 1`] = `
[
  {
    "name": "Wayan Koster",
    "role": "admin",
  },
  {
    "name": "Wayan Koster",
    "role": "admin",
  },
]
`;

exports[`Query authorization ThrowError mode Applied on array of complex type property Should authorize with proper role 2`] = `
{
  "code": "GRAPHQL_AUTHORIZATION_FAILED",
  "paths": [
    "test.role",
  ],
}
`;

exports[`Query authorization ThrowError mode Applied on complex type property Should authorize with multiple directives 1`] = `
{
  "name": "Wayan Koster",
  "role": "admin",
}
`;

exports[`Query authorization ThrowError mode Applied on complex type property Should authorize with multiple directives 2`] = `
{
  "name": "Wayan Koster",
  "role": "admin",
}
`;

exports[`Query authorization ThrowError mode Applied on complex type property Should authorize with multiple directives 3`] = `
{
  "code": "GRAPHQL_AUTHORIZATION_FAILED",
  "paths": [
    "test.role",
  ],
}
`;

exports[`Query authorization ThrowError mode Applied on complex type property Should authorize with proper role 1`] = `
{
  "name": "Wayan Koster",
  "role": "admin",
}
`;

exports[`Query authorization ThrowError mode Applied on complex type property Should authorize with proper role 2`] = `
{
  "code": "GRAPHQL_AUTHORIZATION_FAILED",
  "paths": [
    "test.role",
  ],
}
`;

exports[`Query authorization ThrowError mode Applied on nested complex type property Should authorize with multiple directives 1`] = `
{
  "name": "Wayan Koster",
  "partner": {
    "name": "Wayan Koster",
    "role": "admin",
  },
}
`;

exports[`Query authorization ThrowError mode Applied on nested complex type property Should authorize with multiple directives 2`] = `
{
  "name": "Wayan Koster",
  "partner": {
    "name": "Wayan Koster",
    "role": "admin",
  },
}
`;

exports[`Query authorization ThrowError mode Applied on nested complex type property Should authorize with multiple directives 3`] = `
{
  "code": "GRAPHQL_AUTHORIZATION_FAILED",
  "paths": [
    "test.partner.role",
  ],
}
`;

exports[`Query authorization ThrowError mode Applied on nested complex type property Should authorize with proper role 1`] = `
{
  "name": "Wayan Koster",
  "partner": {
    "name": "Wayan Koster",
    "role": "admin",
  },
}
`;

exports[`Query authorization ThrowError mode Applied on nested complex type property Should authorize with proper role 2`] = `
{
  "code": "GRAPHQL_AUTHORIZATION_FAILED",
  "paths": [
    "test.partner.role",
  ],
}
`;

exports[`Query authorization ThrowError mode Applied on query field with primitive type Should authorize with multiple directives 1`] = `
{
  "code": "GRAPHQL_AUTHORIZATION_FAILED",
  "paths": [
    "test",
  ],
}
`;

exports[`Query authorization ThrowError mode Applied on query field with primitive type Should authorize with proper role 1`] = `
{
  "code": "GRAPHQL_AUTHORIZATION_FAILED",
  "paths": [
    "test",
  ],
}
`;
